### DISCUSSION TITLE
Telit, Sierra Wireless and other modules with USB audio

### DISCUSSION THREAD

**Discussion by IchthysMaranatha** [IchthysMaranatha/asterisk-chan-quectel#11] [Ideas]
With UAC integration done (thanks to God!), it should probably be easy to get other similar modules from other manufacturers to work. If you have a potentially compatible module like Telit LE910 https://sixfab.com/wp-content/uploads/2021/02/Telit_LE910C1_LE910C4_Audio_Over_USB_Application_Note_r2.pdf or Sierra Wireless WP76xx or any other with USB audio feature and are comfortable with using minicom and willing to work through trial and error changes in code, please drop a line here. We could collaborate and try and get your module working with Asterisk.

**Comment by dckmedia** [comment]
I have Quectel EG06-AUTL 4G module extracted from scrap [Arcadyan LH1000](https://retrounit.com.au/products/telstra-smart-modem-gen-2-with-voice-back-up-model-lh1000?variant=43255840309477) modem.

According to [hardware design guide](https://auroraevernet.ru/upload/iblock/334/33423bd08a8e4b8ae8371e4d82b292c2.pdf) it supports voice over USB on Telematics version (page 12)


![Q2](https://user-images.githubusercontent.com/13134612/196833369-a99e2560-d54f-4329-b724-9e81be20750d.JPG)

**Reply by cpanditha** [reply]
@IchthysMaranatha , thank you for pointing me to this Quectel forum. Regarding above module, i have posted a questing to [Quectel forum](https://forums.quectel.com/t/eg06-autl-qdl-mode/18122/2).  

I have connected the EG06-AUTL module to a mini pci slot of a PC this time and got some detection this time when call lsusb.
![8c3f23a97ef64fd90edbf339aefeff94938a3ba2](https://user-images.githubusercontent.com/82849276/200156440-0e87ef82-100b-4af0-a4a6-d92466a7a55f.jpeg)

As per couple of posts in Quectel forum noticed that pin masking worked for some people. So i have tried that and found 3 ports while its is connected to minipci port in that PC.


![ports](https://user-images.githubusercontent.com/82849276/200156591-feebe843-b53e-4c53-b8ae-bb070819c1a7.JPG)

This is the pin i masked to get the 3 ports appear.

![pin](https://user-images.githubusercontent.com/82849276/200156667-d7fc68d7-110a-40ec-b250-653745512ad7.JPG)

But still there is an issue. I'm not getting any detection at all when i connect it to t[his dongle adapter](https://www.ebay.com.au/itm/384923028608).

**Comment by IchthysMaranatha** [comment]
If it has usb voice support, it should work without any modifications to code. Are you getting the ttyusb0-3 ports on linux?

**Reply by cpanditha** [reply]
I have tried connecting this to a laptop mini pci port and tried to boot it but laptop did not boot at all while this module plugged in. May be this only works for the modem itself. I mean the gold pin set is the same size of mini pci slot and fits well. But just noticed that connecting pins from top left corner (P1)looks bit different from normal mini pci slot pins.

**Reply by IchthysMaranatha** [reply]
Laptop may allow only whitelisted components, you could say for sure after connecting it to a router with mpcie slot like ZBT WE826 or Mikrotik RBM33G. Or you could try with a dongle adapter like this https://www.aliexpress.us/item/3256801500256669.html?

**Reply by dckmedia** [reply]
Great, thanks. Found [this one on ebay](https://www.ebay.com.au/itm/384923028608?hash=item599f334c80:g:sqAAAOSwVLhilkkn&amdata=enc%3AAQAHAAAA8M30Vb4eewQ2tt86aMJyBEKt0MHv9NLaBiX9%2BzZboeYTn8fQg3AJ1iAjV%2BXezYthIgK956uFdMD4EmnXteU%2BH%2BBEhUDsOcdthYjbDks58FkEXOwBz6lWfirWkr1UnBk8KkH62lC3HGzCm6Y%2BAjKk8ExfkIc1EPI%2FuK9PXCYP%2BhCgIoBbyUkeap6r1tnuSpzIRi48tbrAQQJvE9%2FiS523zMX2sFAmUehLZ1xT3AsMl%2BmNlVxCD4kGcC5VPEbEh48i4%2F118FhBVldweXKFHItexq7z%2FOeZiOoKI5AxAHWqFUjcO0tEEcoFQDizYIXxKJ0%2B7g%3D%3D%7Ctkp%3ABk9SR5zW5s7-YA) which the seller is local which means can get it faster than Ali. I will try with one of these adapters and let you know.

**Reply by dckmedia** [reply]
With the above EG06-AUTL module and USB dongle adapter not seems to be working. Found another module Quectel EC25-AUX. It works with the dongle. So, im getting ttyusb0-3 ports on EC25-AUX. I have installed Freepbx 16.0.21.9 on a PC compile the module and installation complete. When i go to asterisk console and run `quectel show version`, im getting below.

```
chan_quectel: Channel Driver for Mobile Telephony, Version 1.1, Revision 43gh=47cg
Project Home: https://github.com/bg111/asterisk-chan-quectel/
Bug Reporting: https://github.com/bg111/asterisk-chan-quectel/issues
```

But in asterisk console, im getting following 2 errors.

```
[2022-10-28 06:21:29] ERROR[3780]: chan_quectel.c:277 lock_create: open('/var/lock/LCK..ttyUSB2') failed: Permission denied
[2022-10-28 06:21:29] WARNING[3780]: chan_quectel.c:362 opentty: unable to open /dev/ttyUSB2: Permission denied
```

freepbx*CLI> quectel show devices shows this.

```
ID           Group State      RSSI Mode Submode Provider Name  Model      Firmware          IMEI             IMSI             Number
quectel0     0     Not connec 0    0    0       NONE                                                                          Unknown

```
How can i resolve Permission denied error

**Reply by hellkaim** [reply]
`cat /etc/udev/rules.d/20-voice-usb-modem.rules' 
'# This is used to change the defaultn owner of ttyUSBX for asterisk chan_dongle module
KERNEL=="ttyUSB[0-9]*", OWNER="asterisk", GROUP="dialout", MODE="0660"`

So you use `udev` and instruct it to set permissions after device creation.

The rule might be more precise and you have to change _OWNER_ and _GROUP_ according to your setup.

**Reply by dckmedia** [reply]
Thanks !
```
cat /etc/udev/rules.d/20-voice-usb-modem.rules

Copy and paste:
KERNEL=="ttyUSB*", MODE="0666", OWNER="asterisk", GROUP="uucp"

```
saved the file and rebooted the PC. Error disappeared on freepbx asterisk console. I'm working on the signal/network registration issue as per both your advises. I will let you know.

**Comment by IchthysMaranatha** [comment]
Freepbx runs asterisk as asterisk user. No permission to use ttyusb devices.
Try with
sudo usermod -a -G dialout,lock,uucp asterisk
Please read and follow this to test without freepbx https://github.com/IchthysMaranatha/asterisk-chan-quectel/discussions/13

**Reply by dckmedia** [reply]
Thanks ! As per test bed setup article, i was able to install everything on ubuntu. No more errors. Output of `quectel show devices` as below. State "GSM not re" means its is picking up the signal.?  I'm on Vodafone 4G network and following are the bands they use. B1, B3, B5

-   2100MHz 
-   1800MHz
-   850MHz

When sim inserted in to mobile phone i have 3 bars of signal and show on 4g coverage.

```
ID           Group State      RSSI Mode Submode Provider Name  Model      Firmware          IMEI             IMSI             Number
quectel0     0     GSM not re 99   0    0       NONE           EC25       EC25AUXGAR08A03M1 862738146905502  505008318282320  Unknown
```

**Reply by IchthysMaranatha** [reply]
No, it's not a problem of picking up signal but problem with registering on to the network.
Try this https://forums.quectel.com/t/ec25-eux-volte-support/16038/2
check at+creg after restart, if still not registered, select any one operator from the list and activate that. If still not registered, get help from quectel forum.

**Reply by IchthysMaranatha** [reply]
Unload module from asterisk and use minicom to test and pass commands till registration problem is sorted out.

**Reply by hellkaim** [reply]
> Output of `quectel show devices` as below. State "GSM not re" means its is picking up the signal.? 

I found Quectel devices to take more time to lock to the network compared to say E1550 (classic GSM). I have 3 modems, 2 has 21-24 RSSI and one 99. Why - because. Try to figure out - may be the exact location of your modem is not allowing it to recieve proper signal?

FYI I have another device that has other issue with wireless: if I place it to the location with highest possible RSSI it will loose WiFI (about 30 CM from access point). So I have to move it away like 1-2 meters and WiFi comes back, dropping RSSI to 27-32.

**Reply by dckmedia** [reply]
> Unload module from asterisk and use minicom to test and pass commands till registration problem is sorted out.

I have tried to with minicom by issuing at+creg while module is unloaded. reply was OK. However, reading through Quectel forum found this app called [AT Command Tester for Quectel® Modules](https://m2msupport.net/m2msupport/download-at-command-tester-for-quectel-modules/)

Downloaded[ tool pack](https://www.quectel.com/ProductDownload/EC25.html) from this link and installed the driver for windows and tested the card with USB dongle. It seems to be working as per below. I was able to do following,

- make a test call to my mobile phone
- send an sms to my mobile

![com](https://user-images.githubusercontent.com/13134612/198642133-233d584f-7d20-4273-9d65-755331e71c2b.JPG)

My test bed setup has a wired connection to the ISP internet modem,(about 2 ft apart from ISP modem ) It could be the current location of my device device as @hellkaim mentioned.

**Comment by IchthysMaranatha** [comment]
If at+creg=0,1 then it should show registered when using the module also. That's all it checks for, you should now close all progs accessing modem and try again with the asterisk module.

**Reply by IchthysMaranatha** [reply]
You could move testbed to where you have this Windows machine, just use a vm.

**Reply by dckmedia** [reply]
thanks ! moving the location did the trick.

```
ID           Group State      RSSI Mode Submode Provider Name  Model      Firmware          IMEI             IMSI             Number
quectel0     0     Free       16   0    0       Vodafone AU Le EC25       EC25AUXGAR08A03M1 862738146905502  505008318282320  Unknown
```

Incoming outgoing and SMS works well.

**Reply by IchthysMaranatha** [reply]
Great! If you wish to fine tune, you can try locking bands (from the bands you mentioned band 5 has greatest penetration) in the modem and extend coverage and prevent dropouts
https://forums.quectel.com/t/band-lock-in-ec25-module/11853

**Reply by dckmedia** [reply]
Sure,
 `AT+QCFG="band"` returns  `+QCFG: "band",0x2ff,0x80080000df,0x0`

`AT+QNWINFO` returns `+QNWINFO: "FDD LTE","50503","LTE BAND 3",1550`

I'm struggling to understand how to change it to Band 5.

I have tried `AT+QCFG="band",0,5,0` returns OK and then gave it a restart.  After restart, 
`AT+QNWINFO` now returns `+QNWINFO: "HSPA+","50503","WCDMA 900",3085`

**Reply by dckmedia** [reply]
Update..

It seems `AT+QCFG="band",0,5,0` is the wrong command.

AT+QCFG=“band”,0,10,0 returned OK and now when i do `AT+QNWINFO` it shows the below.
`+QNWINFO: "FDD LTE","50503","LTE BAND 5",2460`

**Reply by IchthysMaranatha** [reply]
Yes, I remember that's how I'd fixed it for band 5 too. Sorry, I just linked to first result in my search
For anyone stumbling upon this, foll values without 0x

0x1 | LTE Band 1 2100MHz
0x4 | LTE Band 3 1800MHz
0x10 | LTE Band 5 850MHz
0x40 | LTE Band 7 2600MHz
0x80 | LTE Band 8 900MHz
0x80000 | LTE Band 20 800MHz
0x7FFFFFFFFFFFFFFF | Any frequency band

**Reply by dckmedia** [reply]
I have tried to setup the UAC on EC25-AUX. I am getting +QCFG="USBCFG",0x2C7C,0x0125,1,1,1,1,1,0,1 for AT+QCFG="USBCFG". I have also done AT+QPCMV=1,2 received OK.


When i issue sudo aplay -L, im getting the following output. 


```
 sudo aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
pulse
    PulseAudio Sound Server
default:CARD=Device
    USB Audio Device, USB Audio
    Default Audio Device
sysdefault:CARD=Device
    USB Audio Device, USB Audio
    Default Audio Device
front:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    Front speakers
surround21:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    Direct sample mixing device
dsnoop:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    Direct sample snooping device
hw:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    Direct hardware device without any conversions
plughw:CARD=Device,DEV=0
    USB Audio Device, USB Audio
    Hardware device with all software conversions
default:CARD=EC25AUX
    EC25-AUX, USB Audio
    Default Audio Device
sysdefault:CARD=EC25AUX
    EC25-AUX, USB Audio
    Default Audio Device
front:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    Front speakers
surround21:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    2.1 Surround output to Front and Subwoofer speakers
surround40:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    4.0 Surround output to Front and Rear speakers
surround41:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    4.1 Surround output to Front, Rear and Subwoofer speakers
surround50:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    5.0 Surround output to Front, Center and Rear speakers
surround51:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    5.1 Surround output to Front, Center, Rear and Subwoofer speakers
surround71:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    7.1 Surround output to Front, Center, Side, Rear and Woofer speakers
iec958:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    IEC958 (S/PDIF) Digital Audio Output
dmix:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    Direct sample mixing device
dsnoop:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    Direct sample snooping device
hw:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    Direct hardware device without any conversions
plughw:CARD=EC25AUX,DEV=0
    EC25-AUX, USB Audio
    Hardware device with all software conversions
hdmi:CARD=PCH,DEV=0
    HDA Intel PCH, HDMI 0
    HDMI Audio Output
hdmi:CARD=PCH,DEV=1
    HDA Intel PCH, HDMI 1
    HDMI Audio Output
hdmi:CARD=PCH,DEV=2
    HDA Intel PCH, HDMI 2
    HDMI Audio Output
hdmi:CARD=PCH,DEV=3
    HDA Intel PCH, HDMI 3
    HDMI Audio Output
hdmi:CARD=PCH,DEV=4
    HDA Intel PCH, HDMI 4
    HDMI Audio Output
dmix:CARD=PCH,DEV=3
    HDA Intel PCH, HDMI 0
    Direct sample mixing device
dmix:CARD=PCH,DEV=7
    HDA Intel PCH, HDMI 1
    Direct sample mixing device
dmix:CARD=PCH,DEV=8
    HDA Intel PCH, HDMI 2
    Direct sample mixing device
dmix:CARD=PCH,DEV=9
    HDA Intel PCH, HDMI 3
    Direct sample mixing device
dmix:CARD=PCH,DEV=10
    HDA Intel PCH, HDMI 4
    Direct sample mixing device
dsnoop:CARD=PCH,DEV=3
    HDA Intel PCH, HDMI 0
    Direct sample snooping device
dsnoop:CARD=PCH,DEV=7
    HDA Intel PCH, HDMI 1
    Direct sample snooping device
dsnoop:CARD=PCH,DEV=8
    HDA Intel PCH, HDMI 2
    Direct sample snooping device
dsnoop:CARD=PCH,DEV=9
    HDA Intel PCH, HDMI 3
    Direct sample snooping device
dsnoop:CARD=PCH,DEV=10
    HDA Intel PCH, HDMI 4
    Direct sample snooping device
hw:CARD=PCH,DEV=3
    HDA Intel PCH, HDMI 0
    Direct hardware device without any conversions
hw:CARD=PCH,DEV=7
    HDA Intel PCH, HDMI 1
    Direct hardware device without any conversions
hw:CARD=PCH,DEV=8
    HDA Intel PCH, HDMI 2
    Direct hardware device without any conversions
hw:CARD=PCH,DEV=9
    HDA Intel PCH, HDMI 3
    Direct hardware device without any conversions
hw:CARD=PCH,DEV=10
    HDA Intel PCH, HDMI 4
    Direct hardware device without any conversions
plughw:CARD=PCH,DEV=3
    HDA Intel PCH, HDMI 0
    Hardware device with all software conversions
plughw:CARD=PCH,DEV=7
    HDA Intel PCH, HDMI 1
    Hardware device with all software conversions
plughw:CARD=PCH,DEV=8
    HDA Intel PCH, HDMI 2
    Hardware device with all software conversions
plughw:CARD=PCH,DEV=9
    HDA Intel PCH, HDMI 3
    Hardware device with all software conversions
plughw:CARD=PCH,DEV=10
    HDA Intel PCH, HDMI 4
    Hardware device with all software conversions

```

2 line in quectel.conf has been set an uncommented as per below.

```
quec_uac=1                             
alsadev=hw:CARD=EC25AUX,DEV=0 
```

But I'm getting below error in asterisk console.

```
    -- [quectel0] Trying to connect on /dev/ttyUSB2...
[2022-11-09 15:59:30] ERROR[4255]: chan_quectel.c:103 alsa_card_init: snd_pcm_open failed: No such device
[2022-11-09 15:59:30] ERROR[4255]: chan_quectel.c:200 soundcard_init: Problem opening ALSA capture device hw:CARD=EC25AUX,DEV=0
    -- [quectel0] Quectel has disconnected
    -- [quectel0] Quectel has connected, initializing...
[2022-11-09 15:59:30] ERROR[4266]: at_queue.c:251 at_queue_run: [quectel0] Error write command 'AT' expected response 'OK' length 3, cancel
[2022-11-09 15:59:30] ERROR[4266]: chan_quectel.c:617 do_monitor_phone: [quectel0] Error adding initialization commands to queue
    -- [quectel0] Error initializing Quectel
    -- [quectel0] Quectel has disconnected
```

**Reply by IchthysMaranatha** [reply]
Pulseaudio may be preventing access to asterisk. Might be a user permissions issue.
https://forums.gentoo.org/viewtopic-t-1031386-start-0.html
Disable pulseaudio or you could try this maybe
https://askubuntu.com/questions/666724/how-to-exclude-certain-device-from-being-grabbed-by-pulseaudio

**Reply by dckmedia** [reply]
Finally figured this out. Problem was running asterisk as asterisk user. When I started running asterisk as root, any of the above errors such as, `chan_quectel.c:103 alsa_card_init: snd_pcm_open failed: No such device` has stopped. Not ideal to run asterisk as root i believe. But this is just for testing UAC. So confirming it works great. I felt call quality has been lot improved after this. 

Also faced a bit of a struggle after this as dongle started disconnected time to time. Asterisk console started showing failed and timed out AT commands. Initially i thought it may be a power supply issue. But later found out that somehow data mode of the card had turned on (may be i tried the wrong AT command at some point) and Modem manager in ubuntu was also using the card simultaneously. So, had to remove it total and it is now stable.

**Comment by IchthysMaranatha** [comment]
@cpanditha
Have you tested if voice calls are working on linux on the PC (with asterisk) after the masking and getting the ports available? You could troubleshoot getting it to work with the adapter once you're sure that it's suitable for calls.

**Reply by cpanditha** [reply]
This particular module doesn't have a dedicated sim slot behind it. I have install asterisk to the PC with the mini pci slot. But unable to make test call. I can successfully connect to the module using minicom. Are there any special AT command that i can try to make sure that it supports voice.

![back](https://user-images.githubusercontent.com/82849276/200162814-9db348f3-4ceb-4a95-add7-6134619be4e9.png)

**Comment by IchthysMaranatha** [comment]
We won't know for sure without an active sim card, commands to check will not run without it. 
Maybe you've already gone through these
https://forums.quectel.com/t/ec21-eux-minipcie-module-enters-qualcom-qdl-mode/12255
https://forums.quectel.com/t/ep06-e-over-usb-identified-as-05c6-9008/3851/13
You might have to add pull-up resistors if it doesn't work by masking alone
